File: Week_12/Practicum/Find if Path Exists in Graph.py
def find(x, parents):
    if parents[x] == x:
        return x
    
    furthest_parent = find(parents[x], parents)
    parents[x] = furthest_parent

    return furthest_parent

def union(x, y, parents):
    x_root = find(x, parents)
    y_root = find(y, parents)

    if x_root != y_root:
        parents[x_root] = y_root
        return 1
        
    return 0

class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        parents = [i for i in range(n)]
        for edge in edges:
            union(edge[0], edge[1], parents)

        return find(source, parents) == find(destination, parents)

File: Week_12/Practicum/Find if Path Exists in Graph.cpp
class Solution {

struct DisjointSet {
    vector<int> parents;

    DisjointSet(int n) {
        for (int i =0; i < n; i++) {
            parents.push_back(i);
        }
    }

    int find(int v) {
        if (v == parents[v])
            return v;
        return parents[v] = find(parents[v]);
    }

    void unite(int u, int v) {
        int parentU = find(u);
        int parentV = find(v);

        if (parentU == parentV) {
            return;
        }

        parents[parentU] = parentV;
    }
};

public:
    bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
        DisjointSet ds(n);

        for (auto& edge : edges) {
            ds.unite(edge[0], edge[1]);
        }

        return ds.find(source) == ds.find(destination);
    }
};